### PR TITLE
Solver stability improvements

### PR DESCRIPTION
**Pull request by richardotis**
Note: This PR depends on #82.

This PR adds a few stability fixes to the solver:
1. Keep track of previously removed CompositionSets and prevent them from being re-added.
2. If the active set of phases has not changed, just calculate the change in chemical potentials for that iteration, rather than computing the full chemical potentials. This seems to confer a stability improvement.
3. Remove the unwieldy `obj_weight` scheme, which was full of magic numbers and didn't seem to work reliably.

~~I have marked this as WIP since there's some obsolete code related to `obj_weight` that still needs to be removed, but it's otherwise in shape for review.~~

### CODE REVIEW

**Comment by richardotis** [comment]
Cool fact: For the first time, the phase diagrams of Al-Zn and Al-Fe in BinaryExamples will calculate with zero convergence failures. (I have not yet checked the chemical potentials to make sure, e.g., #76 is fixed.) For Al-Ni, I am still getting a singular matrix crash (which I suspect is related to numerical instability from magnetism), but all the other convergence failures are gone.

File: pycalphad/core/eqsolver.pyx [hunk 1]
```diff
@@ -424,13 +448,20 @@ def _solve_eq_at_conditions(comps, properties, phase_records, grid, conds_keys,
             master_hess[:num_vars, num_vars:] = -constraint_jac.T
             master_hess[num_vars:, :num_vars] = constraint_jac
             master_grad = np.zeros(l_hessian.shape[0] + l_constraints.shape[0])
-            master_grad[:l_hessian.shape[0]] = -np.array(gradient_term)
+            master_grad[:l_hessian.shape[0]] = -np.array(gradient_term) + np.dot(constraint_jac.T, l_multipliers)
```
Comment (bocklund) [inline pycalphad/core/eqsolver.pyx]: Was this an implementation error or what is this?

File: pycalphad/core/eqsolver.pyx [hunk 2]
```diff
@@ -8,7 +8,7 @@ import scipy.spatial
 import collections
 from pycalphad.core.composition_set cimport CompositionSet
 from pycalphad.core.phase_rec cimport PhaseRecord, PhaseRecord_from_f2py
-from pycalphad.core.constants import MIN_SITE_FRACTION, COMP_DIFFERENCE_TOL, BIGNUM
+from pycalphad.core.constants import MIN_SITE_FRACTION, MIN_PHASE_FRACTION, MIN_PHASE_FRACTION, COMP_DIFFERENCE_TOL, BIGNUM
```
Comment (bocklund) [inline pycalphad/core/eqsolver.pyx]: `MIN_PHASE_FRACTION` is imported twice

**Comment by richardotis** [comment]
@bocklund This is ready for another look.

**Comment by bocklund** [comment]
LGTM

**Pull request merged**